Framework/Algorithms/src/CropWorkspaceRagged.cpp

```diff
@@ -0,0 +1,166 @@
+#include "MantidKernel/BoundedValidator.h"
+#include "MantidKernel/MandatoryValidator.h"
+
+namespace {
+std::vector<double> getSubVector(Mantid::MantidVec &data,
+                                 const int64_t &lowerIndex,
+                                 const int64_t &upperIndex) {
+  auto low = std::next(data.begin(), lowerIndex);
+  auto up = std::next(data.begin(), upperIndex);
+  // get new vectors
+  std::vector<double> newData(low, up);
+  return newData;
+}
+
+} // namespace
+
+namespace Mantid {
+namespace Algorithms {
+
+using namespace Kernel;
+using namespace API;
+
+DECLARE_ALGORITHM(CropWorkspaceRagged)
+
+/// Init function
+void CropWorkspaceRagged::init() {
+  declareProperty(std::make_unique<WorkspaceProperty<MatrixWorkspace>>(
+                      "InputWorkspace", "", Direction::Input),
+                  "The input workspace");
+  declareProperty(std::make_unique<WorkspaceProperty<MatrixWorkspace>>(
+                      "OutputWorkspace", "", Direction::Output),
+                  "Name to be given to the cropped workspace.");
+
+  auto required = std::make_shared<MandatoryValidator<std::vector<double>>>();
+  declareProperty(std::make_unique<ArrayProperty<double>>("XMin", required),
+                  "The value(s) to start the cropping from. Should be either a "
+                  "single value or a list.");
+  declareProperty(std::make_unique<ArrayProperty<double>>("XMax", required),
+                  "The value(s) to end the cropping at. Should be either a "
+                  "single value or a list.");
+}
+
+/// Input validation
+std::map<std::string, std::string> CropWorkspaceRagged::validateInputs() {
+  std::map<std::string, std::string> issues;
+  MatrixWorkspace_sptr ws = getProperty("InputWorkspace");
+  auto numSpectra = ws->getNumberHistograms();
+  std::vector<double> xMin = getProperty("XMin");
+  std::vector<double> xMax = getProperty("XMax");
+  if (xMin.size() == 0 || (xMin.size() != numSpectra && xMin.size() > 1)) {
+    issues["XMin"] = "XMin must be a single value or one value per sepctrum.";
+  }
+  if (xMax.size() == 0 || (xMax.size() > 1 && xMax.size() != numSpectra)) {
+    issues["XMax"] = "XMax must be a single value or one value per sepctrum.";
+  }
+  if (xMin.size() == 1 && xMax.size() == 1 && xMin[0] > xMax[0]) {
+    issues["XMax"] = "XMax must be greater than XMin.";
+  } else if (xMin.size() == 1 && xMax.size() > 1) {
+    for (auto max : xMax) {
+      if (max < xMin[0]) {
+        issues["XMax"] = "XMax must be greater than XMin.";
+        return issues;
+      }
+    }
+  } else if (xMin.size() > 1 && xMax.size() == 1) {
+    for (auto min : xMin) {
+      if (min > xMax[0]) {
+        issues["XMin"] = "XMin must be less than XMax.";
+        return issues;
+      }
+    }
+  } else if (xMin.size() > 1 && xMax.size() > 1) {
+    for (size_t k = 0; k < xMin.size(); k++) {
+      if (xMin[k] > xMax[k]) {
+        issues["XMin"] = "XMin must be less than XMax.";
```

Comment:
Since we have the index is it worth adding the index to the error message. If the index is used it would also make more sense to continue looping and add each failed index rather than return straight away.

Reply:
Realistically, its not that helpful in terms of feedback. If you have 100 spectra (thats a small number) knowing its index 56 still means checking everything. Realistically no one will count the indices to get to the right one.